src/otx/core/model/entity/visual_prompting.py

```diff
@@ -30,46 +49,920 @@ class OTXVisualPromptingModel(
+
+        ## get matched indices
+        matched_matrix = np.expand_dims(idx_grid, axis=-1) == idx_grid_unique  # (totalN, uniqueN)
+
+        ## sample fg_coords_scores matched by matched_matrix
+        matched_grid = np.expand_dims(fg_coords_scores, axis=1) * np.expand_dims(matched_matrix, axis=-1)
+
+        ## sample the highest score one of the samples that are in the same grid
+        matched_indices = self._topk_numpy(matched_grid[..., -1], k=1, axis=0, largest=True)[1][0].astype(np.int64)
+        points_scores = matched_grid[matched_indices].diagonal().T
+
+        ## sort by the highest score
+        sorted_points_scores_indices = np.flip(np.argsort(points_scores[:, -1]), axis=-1).astype(np.int64)
+        points_scores = points_scores[sorted_points_scores_indices]
+
+        # Top-last point selection
+        bg_indices = self._topk_numpy(mask_sim.flatten(), num_bg_points, largest=False)[1]
+        bg_x = np.expand_dims(bg_indices // w_sim, axis=0)
+        bg_y = bg_indices - bg_x * w_sim
+        bg_coords = np.concatenate((bg_y, bg_x), axis=0).transpose(1, 0)
+        bg_coords = bg_coords.astype(np.float32)
+
+        return points_scores, bg_coords
+
+    def _resize_to_original_shape(self, masks: np.ndarray, image_size: int, original_shape: np.ndarray) -> np.ndarray:
+        """Resize feature size to original shape."""
+        # resize feature size to input size
+        masks = cv2.resize(masks, (image_size, image_size), interpolation=cv2.INTER_LINEAR)
+
+        # remove pad
+        prepadded_size = self._get_prepadded_size(original_shape, image_size)
+        masks = masks[..., : prepadded_size[0], : prepadded_size[1]]
+
+        # resize unpadded one to original shape
+        original_shape = original_shape.astype(np.int64)
+        h, w = original_shape[0], original_shape[1]
+        return cv2.resize(masks, (w, h), interpolation=cv2.INTER_LINEAR)
+
+    def _get_prepadded_size(self, original_shape: int, image_size: int) -> np.ndarray:
+        """Get pre-padded size."""
+        scale = image_size / np.max(original_shape)
+        transformed_size = scale * original_shape
+        return np.floor(transformed_size + 0.5).astype(np.int64)
+
+    def _inspect_overlapping_areas(
+        self,
+        predicted_masks: dict[int, list[np.ndarray]],
+        used_points: dict[int, list[np.ndarray]],
+        threshold_iou: float = 0.8,
+    ) -> None:
+        def _calculate_mask_iou(mask1: np.ndarray, mask2: np.ndarray) -> np.ndarray:
+            assert mask1.ndim == 2  # noqa: S101
+            assert mask2.ndim == 2  # noqa: S101
+            intersection = np.logical_and(mask1, mask2).sum().item()
+            union = np.logical_or(mask1, mask2).sum().item()
+
+            # Avoid division by zero
+            if union == 0:
+                return 0.0
+            return intersection / union
+
+        for (label, masks), (other_label, other_masks) in product(predicted_masks.items(), predicted_masks.items()):
```

Comment:
Is this typo? `product(predicted_masks.items(), predicted_masks.items())`

Reply:
No, it is to check all labeled masks. It will be skipped if `other_label <= label` at the next line.